src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Cache/CacheAspectThreadManagementIntegrationTests.cs

```diff
@@ -0,0 +1,599 @@
+            events[2].AssertEventMessageEquals($"{nameof(AsyncActionFilter)} OnActionExecutionAsync completed - returned Task IsCompleted = False");
+        
+        // We expect this to be called almost immediately after the cached item request call.
+        Assert.InRange(asyncActionFilterCompleted.Millis, requestingCachedItem.Millis, requestingCachedItem.Millis + noDelayTimeMillis);
+
+        // 5. Because the cache finds an existing item with the given key, it's able to return that in its continuation
+        //    and the AsyncResultFilter is then able to return that with the HTTP response.
+        var existingCachedItemResponse = 
+            events[3].AssertEventMessageEquals($"Returning cached item {expectedControllerResult} with key {expectedCacheKey} after delay");
+        
+        // We expect this to be called after the delay time that it took to check for a cached item.
+        Assert.InRange(
+            existingCachedItemResponse.Millis, 
+            requestingCachedItem.Millis + AsyncOperationDelayMillis - TaskDelayErrorToleranceMillis, 
+            (requestingCachedItem.Millis + AsyncOperationDelayMillis * 2) - 1);
+
+        var asyncResultFilterCalled =
+            events[4].AssertEventMessageEquals($"{nameof(AsyncResultFilter)} OnResultExecutionAsync called");
+        
+        // We expect this to have been run just after the existing cached item is returned as the result from the
+        // Controller method to the MVC framework.
+        Assert.InRange(
+            asyncResultFilterCalled.Millis, 
+            existingCachedItemResponse.Millis, 
+            existingCachedItemResponse.Millis + noDelayTimeMillis);
+
+        var asyncResultFilterCompleted =
+            events[5].AssertEventMessageEquals($"{nameof(AsyncResultFilter)} OnResultExecutionAsync completed - returned Task IsCompleted = True");
+        
+        // We expect this to have been run just after the response is provided from the Controller.
+        Assert.InRange(
+            asyncResultFilterCompleted.Millis, 
+            asyncResultFilterCalled.Millis, 
+            asyncResultFilterCalled.Millis + noDelayTimeMillis);
+    }
+    
+    /// <summary>
+    /// This test simulates creating a number of simultaneous HTTP requests and measuring the effects in terms of
+    /// time to process them all and the growth of the Worker Thread pool during the process.
+    ///
+    /// In a scenario where there was blocking code accidentally introduced, the current Worker Thread count would
+    /// immediately increase as all the available threads get blocked. The time taken to complete processing all of
+    /// the requests would also greatly exceed the expected maximum time to complete. This has been proven by
+    /// deliberately adding blocking back into the <see cref="CacheAttribute.WrapAsync{T}"/> implementation and seeing
+    /// these tests fail.
+    /// </summary>
+    /// <exception cref="Exception"></exception>
+    [Fact]
```

Comment:
Shouldn't this have a timeout to avoid hanging indefinitely if the loop never ends?

Reply:
(that's where the `<exception cref="Exception"></exception> came from in the method doco, which I will also embellish upon)